### PR TITLE
Updates the Builders to have a Validate Method

### PR DESCRIPTION
**Pull request by jeffreyladd**
# Summary
Currently, the Edit Webhook Message does not validate that Files were not attached to the Builder.  We need to Validate that and return an error before we send the message to the API or else users will think they can send files in an edit message even though they cant.

### CODE REVIEW

File: DSharpPlus/Entities/DiscordWebhook.cs [hunk 1]
```diff
@@ -137,8 +137,13 @@ public Task ExecuteGithubAsync(string json)
         /// <exception cref="Exceptions.NotFoundException">Thrown when the webhook does not exist.</exception>
         /// <exception cref="Exceptions.BadRequestException">Thrown when an invalid parameter was provided.</exception>
         /// <exception cref="Exceptions.ServerErrorException">Thrown when Discord is unable to process the request.</exception>
-        public Task<DiscordMessage> EditMessageAsync(ulong messageId, DiscordWebhookBuilder builder)
-            => (this.Discord?.ApiClient ?? this.ApiClient).EditWebhookMessageAsync(this.Id, this.Token, messageId, builder);
+        public async Task<DiscordMessage> EditMessageAsync(ulong messageId, DiscordWebhookBuilder builder)
+        {
+            if (builder.Files.Any())
+                throw new ArgumentException("You cannot add files when modifing a message.");
```
Comment (Neuheit) [inline DSharpPlus/Entities/DiscordWebhook.cs:143]:
```suggestion
                throw new ArgumentException("You cannot add files when modifying a message.");
```

File: DSharpPlus.Rest/DiscordRestClient.cs [hunk 2]
```diff
@@ -1225,8 +1225,13 @@ public Task<DiscordMessage> ExecuteWebhookAsync(ulong webhook_id, string webhook
         /// <param name="messageId">The id of the message to edit.</param>
         /// <param name="builder">The builder of the message to edit.</param>
         /// <returns>The modified <see cref="DiscordMessage"/></returns>
-        public Task<DiscordMessage> EditWebhookMessageAsync(ulong webhook_id, string webhook_token, ulong messageId, DiscordWebhookBuilder builder)
-            => this.ApiClient.EditWebhookMessageAsync(webhook_id, webhook_token, messageId, builder);
+        public async Task<DiscordMessage> EditWebhookMessageAsync(ulong webhook_id, string webhook_token, ulong messageId, DiscordWebhookBuilder builder)
+        {
+            if (builder.Files.Any())
+                throw new ArgumentException("You cannot add files when modifing a message.");
```
Comment (Neuheit) [inline DSharpPlus.Rest/DiscordRestClient.cs:1231]:
```suggestion
                throw new ArgumentException("You cannot add files when modifying a message.");
```

**Comment by IDoEverything** [comment]
If there should be checks for unusable fields while modifying messages, shouldn't we also add checks for replies in regular message editing, and avatar and username for webhooks as well?

**Comment by jeffreyladd** [comment]
> If there should be checks for unusable fields while modifying messages, shouldn't we also add checks for replies in regular message editing, and avatar and username for webhooks as well?

Yes, u are 100% correct (I missed the replies one on the message builder).  I have updated the builders to have a validate method with the proper validation and altered the internal code to call this method.

File: DSharpPlus/Entities/DiscordMessageBuilder.cs [hunk 3]
```diff
@@ -239,5 +239,26 @@ public void Clear()
...
+        /// <summary>
+        /// Does the validation before we send a the Create/Modify request.
+        /// </summary>
+        /// <param name="isModify">Tells the method to perform the Modify Validation or Create Validation.</param>
+        internal void Validate(bool isModify = false)
+        {
+            if (isModify)
+            {
+                if (this.Files.Any())
+                    throw new ArgumentException("You cannot add files when modifing a message.");
```
Comment (Neuheit) [inline DSharpPlus/Entities/DiscordMessageBuilder.cs:252]:
```suggestion
                    throw new ArgumentException("You cannot add files when modifying a message.");
```

File: DSharpPlus/Entities/DiscordWebhookBuilder.cs [hunk 4]
```diff
@@ -261,5 +261,29 @@ public void Clear()
...
+        /// <param name="isModify">Tells the method to perform the Modify Validation or Create Validation.</param>
+        internal void Validate(bool isModify = false)
+        {
+            if (isModify)
+            {
+                if (this.Files.Any())
+                    throw new ArgumentException("You cannot add files when modifying a message.");
+
+                if (this.Username.HasValue)
+                    throw new ArgumentException("You cannot change the Username of a message.");
```
Comment (Neuheit) [inline DSharpPlus/Entities/DiscordWebhookBuilder.cs:277]:
```suggestion
                    throw new ArgumentException("You cannot change the username of a message.");
```

File: DSharpPlus/Entities/DiscordWebhookBuilder.cs [hunk 5]
```diff
@@ -261,5 +261,29 @@ public void Clear()
...
+
+                if (this.AvatarUrl.HasValue)
+                    throw new ArgumentException("You cannot change the Avatar of a message.");
```
Comment (Neuheit) [inline DSharpPlus/Entities/DiscordWebhookBuilder.cs:280]:
```suggestion
                    throw new ArgumentException("You cannot change the avatar of a message.");
```

File: DSharpPlus/Net/DiscordApiClient.cs [hunk 6]
```diff
@@ -774,17 +774,7 @@ internal async Task<DiscordMessage> CreateMessageAsync(ulong channel_id, string
 
         internal async Task<DiscordMessage> CreateMessageAsync(ulong channel_id, DiscordMessageBuilder builder)
         {
-            if (builder.Content != null && builder.Content.Length > 2000)
```
Comment (Neuheit) [inline DSharpPlus/Net/DiscordApiClient.cs:777]: You remove these checks here but don't specify them in the builder validate method. I would copy these checks and place them in the validate `else` statement (where it doesn't modify).

Comment (jeffreyladd) [reply DSharpPlus/Net/DiscordApiClient.cs:777]: The Content Prop of the message builder already does this check for line 777.

File: DSharpPlus/Entities/DiscordWebhookBuilder.cs [hunk 7]
```diff
@@ -261,5 +261,29 @@ public void Clear()
...
+
+                if (this.Username.HasValue)
+                    throw new ArgumentException("You cannot change the username of a message.");
+
+                if (this.AvatarUrl.HasValue)
+                    throw new ArgumentException("You cannot change the avatar of a message.");
+            }
+            else
+            {
+                if (this.Files?.Count == 0 && string.IsNullOrEmpty(this.Content) && this.Embeds == null)
```
Comment (Neuheit) [inline DSharpPlus/Entities/DiscordWebhookBuilder.cs:284]: `this.Embeds` will never be null, you should check for the collection length instead like you do for the files.

**Review by Neuheit — APPROVED** [review]

**Pull request merged**